### PR TITLE
Cooking Overhaul

### PR DESCRIPTION
**Pull request by jellysnake**
# Details
Redo the Bush and Vine growing systems.
There is now a `SeedComponent` that allows an item to be planted. The same prefab is used for vine seed and for bush seeds.  
Depending on the fields in this component a vine or a bush entity will be created at that spot.  

### Bushes
For bush entities they will cycle through the blocks defined in the `growthStages` field of the `BushDefinition` component on the prefab. Once the last stage is reached it will be harvestable and the produce will be given to the player. The block will then regress back one stage or be destroyed.  
When a bush is destroyed a seed will drop. Additional seeds will drop if the bush was fully grown.

### Vines
Vine entity will simply create a snaking path of the block given in `stem` field of the `VineDefinition`. At points along it's path it will create instances of the entity given by the `bud` field. These buds act exactly the same as the bushes above. When a stem block in the bush is broken, all other blocks down of the stem from that point will be destroyed. If the stem is the destroyed block then the entire vine will be destroyed.

All the produce has been converted over to the new format. They will be removed and split into content packs at some point in the future.

# Testing
Simply spawn in any produce and verify it behaves as above

### CODE REVIEW

**Comment by jellysnake** [comment]
@Cervator this is now ready to be merged.

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+     * Called when the bush should grow
+     *
+     * @param event         The event indicating the timer has ended
+     * @param bush          The bush to grow
+     * @param bushComponent The bush's definition
+     */
+    @ReceiveEvent
+    public void onBushGrowth(DelayedActionTriggeredEvent event, EntityRef bush, BushDefinitionComponent bushComponent) {
+        EntityRef newBush = null;
+        if (bushComponent.currentStage < bushComponent.stages.length - 1) {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: good thing we have `isInLastStage(bushComponent)`, so we can replace this with `!isInLastStage(bushComponent)`

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+            newBush = doBushGrowth(bushComponent, 1);
+        }
+        resetDelay(newBush == null ? bush : newBush,
+                bushComponent.stages[bushComponent.currentStage].minTime,
+                bushComponent.stages[bushComponent.currentStage].maxTime);
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: i see a method `getCurrentStage(bushComponent)` or a helper method `bushComponent.getCurrentStage()` here :)

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    /**
+     * Called when an attempt to harvest the bush is made.
+     * If the bush is
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: This sentence looks like it is not finished which makes

;)

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+     * @param event  The activation event
+     * @param entity The entity doing the harvesting
+     */
+    @ReceiveEvent
+    public void onHarvest(ActivateEvent event, EntityRef entity) {
+        EntityRef target = event.getTarget();
+        EntityRef harvester = entity.equals(target) ? event.getInstigator() : entity;
+        /* Ensure the target is a plant and the entities are valid */
+        if (!event.isConsumed() && target.exists() && harvester.exists()
+                && target.hasComponent(BushDefinitionComponent.class)) {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: something like  `if(!event.isConsumed() && isValidPlantHarvest(target,harvester))`?

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            BushDefinitionComponent bushComponent = target.getComponent(BushDefinitionComponent.class);
+
+            /* Produce is only given in the final stage */
+            if (bushComponent.currentStage == bushComponent.stages.length - 1) {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: we now have a method for this =)

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+                EntityRef produce;
+                /* Handle a dodgy produce */
+                try {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: extract everything from the try block to a `produce(...)` method.

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            worldProvider.setBlock(bushComponent.position, blockManager.getBlock(BlockManager.AIR_ID));
+        }
+
+        /* Drop some seeds */
+        for (int i = 0; i < numSeeds; i++) {
+            Prefab seed = bushComponent.seed == null ? bushComponent.produce : bushComponent.seed;
+            EntityRef seedItem = entityManager.create(seed);
+            Vector3f position = bushComponent.position.toVector3f().add(0, 0.5f, 0);
+            seedItem.send(new DropItemEvent(position));
+            seedItem.send(new ImpulseEvent(random.nextVector3f(30.0f)));
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: extract the magic number to a constant like `DROP_SEED_IMPULSE`

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+            if (bushComponent.currentStage == bushComponent.stages.length - 1) {
+                EntityRef produce;
+                /* Handle a dodgy produce */
+                try {
+                    produce = entityManager.create(bushComponent.produce);
+                    boolean giveSuccess = inventoryManager.giveItem(harvester, target, produce);
+                    if (!giveSuccess) {
+                        Vector3f position = event.getTargetLocation().add(0, 0.5f, 0);
+                        produce.send(new DropItemEvent(position));
+                        produce.send(new ImpulseEvent(random.nextVector3f(15.0f)));
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: extract the magic number to something like `DROP_PRODUCE_IMPULSE`

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+                    }
+                } catch (NullPointerException e) {
+                    logger.error("Unable to create produce: " + bushComponent.produce);
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: don't catch nullpointerexceptions, do a null check if needed (logging in the else-case is okay though).

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        event.consume();
+    }
+
+    /**
+     * Grow the bush by one stage in a specified direction
+     *
+     * @param bushComponent The definition of the bush to grow
+     * @param direction     The direction to grow in. positive indicates forward & negative indicated backwards
+     */
+    private EntityRef doBushGrowth(BushDefinitionComponent bushComponent, int direction) {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: The term direction is already used for a 3d-direction. I would rename direction to offset / stageOffset

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+        Object[] values = growthStages.values().toArray();
+        Iterator<String> keys = growthStages.keySet().iterator();
+        GrowthStage[] stages = new GrowthStage[values.length];
+        for (int i = 0; i < values.length; i++) {
+            String block = keys.next();
+            stages[i] = new GrowthStage((GrowthStage) values[i]);
+            try {
+                stages[i].block = blockManager.getBlock(block);
+            } catch (NullPointerException e) {
+                logger.error("Unable to get block: " + block);
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: dont catch NPE here

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+     * Builds the list of growth stages from the prefab data
+     *
+     * @param growthStages The prefab GrowthStage data
+     * @return The array of growth stages
+     */
+    private GrowthStage[] buildGrowthStages(Map<String, GrowthStage> growthStages) {
+        Object[] values = growthStages.values().toArray();
+        Iterator<String> keys = growthStages.keySet().iterator();
+        GrowthStage[] stages = new GrowthStage[values.length];
+        for (int i = 0; i < values.length; i++) {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: have a look a `Map.entrySet() for this kind of iteration, you can access key and value from the same object if you iterate over entries`.

Also this looks like it depends on the order of the map which is (likely) not guaranteed for persisted components, better double check it here =)

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    /**
+     * Adds a new delay of random length between the items
+     *
+     * @param entity The entity to have the timer set on
+     * @param min    The minimum duration
+     * @param max    THe maximum duration
+     */
+
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: floating blank line

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+            stages[i].block.setKeepActive(true);
+        }
+        return stages;
+    }
+
+    /**
+     * Adds a new delay of random length between the items
+     *
+     * @param entity The entity to have the timer set on
+     * @param min    The minimum duration
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: duration unit missing (seconds?)

File: src/main/java/org/terasology/simpleFarming/systems/PlantAuthoritySystem.java [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+                return;
+            }
+            plantEntity.send(new OnSeedPlanted(position));
+            inventoryManager.removeItem(seed.getOwner(), seed, seed, true, 1);
+            event.consume();
+        }
+    }
+
+    private boolean checkPosition(Vector3i position) {
+        if (position == null) {
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/PlantAuthoritySystem.java]: rename to `isAirOnTopOfInpenetrableBlock(position)`

File: src/main/java/org/terasology/simpleFarming/systems/PlantAuthoritySystem.java [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+            return false;
+        }
+        Block targetBlock = worldProvider.getBlock(position);
+        if (targetBlock != airBlock) {
+            return false;
+        }
+        targetBlock = worldProvider.getBlock(new Vector3i(position.x(), position.y() - 1, position.z()));
```
Comment (oniatus) [inline src/main/java/org/terasology/simpleFarming/systems/PlantAuthoritySystem.java]: This is not the target block, use a new variable (they are cheap here) like `belowTargetBlock`

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java (hunk 10 above)
Comment (jellysnake) [reply src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: Whilst direction can be used for a 3-d direction there isn't an instance of it being used like that in this module so I don't think it's confusing.

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java (hunk 12 above)
Comment (jellysnake) [reply src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: Order seems to be guaranteed for persisted components as all of my testing hasn't turned up an instance where it isn't.

**Comment by jellysnake** [comment]
Thanks for the review!

I've made most of the changes you've outlined and I went over and cleaned up all the prefab/block files as well.

**Comment by Cervator** [comment]
Nice work @jellysnake & nice review @oniatus - much appreciated :-)

Is this entirely independent from https://github.com/Terasology/Cooking/pull/126 and any other parts of the cooking overhaul?

**Comment by jellysnake** [comment]
Yup, the way the overhaul is structured each step builds on the previous.
If you would rather merge it later that's fine, but the next step will build on this. 
(The next step being to split content off into modules and begin cooking integration.)

**Comment by Cervator** [comment]
Tested it out and am ready to merge if my understanding of things is correct - left a few notes on Slack :-)

Noted two things: a few missing textures from vine-based produce (probably ends up somewhere else) + I'm not sure where the old apple tree, peach tree, etc are hiding

**Pull request merged**